narwhal/primary/src/primary.rs

```diff
@@ -547,6 +542,223 @@ impl PrimaryReceiverHandler {
+                ))
+            })?;
+        ensure!(
+            header.author == *peer_authority,
+            DagError::NetworkError(format!(
+                "Header author {:?} must match requesting peer {peer_authority:?}",
+                header.author
+            ))
+        );
+
+        debug!(
+            "Processing vote request for {:?} round:{:?}",
+            header, header.round
+        );
+
+        // Clone the round updates channel so we can get update notifications specific to
+        // this RPC handler.
+        let mut rx_narwhal_round_updates = self.rx_narwhal_round_updates.clone();
+        let mut narwhal_round = *rx_narwhal_round_updates.borrow();
+        ensure!(
+            narwhal_round <= header.round,
+            DagError::TooOld(header.digest().into(), header.round, narwhal_round)
+        );
+
+        // If requester has provided us with parent certificates, process them all
+        // before proceeding. This may advance our round, so do it before checking round.
+        let mut notifies = Vec::new();
+        for certificate in request.body().parents.clone() {
+            let (tx_notify, rx_notify) = oneshot::channel();
+            notifies.push(rx_notify);
+            self.tx_certificates
+                .send((certificate, Some(tx_notify)))
+                .await
+                .map_err(|_| DagError::ChannelFull)?;
+        }
+        let mut wait_notifies = futures::future::try_join_all(notifies);
+        loop {
+            tokio::select! {
+                results = &mut wait_notifies => {
+                    let results: Result<Vec<_>, _> = results
+                        .map_err(|e| DagError::ClosedChannel(format!("{e:?}")))?
+                        .into_iter()
+                        .collect();
+                    results?;
+                    break
+                },
+                result = rx_narwhal_round_updates.changed() => {
+                    result.unwrap();
+                    narwhal_round = *rx_narwhal_round_updates.borrow();
+                    ensure!(
+                        narwhal_round <= header.round,
+                        DagError::TooOld(header.digest().into(), header.round, narwhal_round)
+                    )
+                },
+            }
+        }
+
+        // Ensure we have the parents. If any are missing, the requester should provide them on retry.
+        let (parents, missing) = self.synchronizer.get_parents(header)?;
+        if !missing.is_empty() {
+            return Ok(RequestVoteResponse {
+                vote: None,
+                missing,
+            });
+        }
+
+        // Now that we've got all the required certificates, ensure we're voting on a
+        // current Header.
+        narwhal_round = *rx_narwhal_round_updates.borrow();
+        ensure!(
+            narwhal_round <= header.round,
+            DagError::TooOld(header.digest().into(), header.round, narwhal_round)
+        );
```

Comment:
We do, because there's no guarantee the select branch on 622 will ever run. The check at the beginning isn't necessary for correctness but just to be "nice" and exit early before processing provided certs if it's going to fail regardless.